File: Raven.Json/Linq/RavenJObject.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Raven.Json.Utilities;

namespace Raven.Json.Linq
{
    public class RavenJObject : RavenJToken
    {
        /// <summary>
        /// Gets the node type for this <see cref="RavenJToken"/>.
        /// </summary>
        /// <value>The type.</value>
        public override JTokenType Type
        {
            get { return JTokenType.Object; }
        }

        public IDictionary<string, RavenJToken> Properties
        {
            get { return _properties ?? (_properties = new CopyOnWriteJDictionary<string>()); }
        }
        private CopyOnWriteJDictionary<string> _properties;

        /// <summary>
        /// Initializes a new instance of the <see cref="RavenJObject"/> class.
        /// </summary>
        public RavenJObject()
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="RavenJObject"/> class from another <see cref="RavenJObject"/> object.
        /// </summary>
        /// <param name="other">A <see cref="RavenJObject"/> object to copy from.</param>
        public RavenJObject(RavenJObject other)
        {
            _properties = (CopyOnWriteJDictionary<string>) other._properties.Clone();
        }

        public override RavenJToken CloneToken()
        {
            return new RavenJObject(this);
        }

        /// <summary>
        /// Creates a <see cref="RavenJObject"/> from an object.
        /// </summary>
		/// <param name="o">The object that will be used to create <see cref="RavenJObject"/>.</param>
		/// <returns>A <see cref="RavenJObject"/> with the values of the specified object</returns>
        public static new RavenJObject FromObject(object o)
        {
            return FromObject(o, new JsonSerializer());
        }

        /// <summary>
        /// Creates a <see cref="RavenJArray"/> from an object.
        /// </summary>
        /// <param name="o">The object that will be used to create <see cref="RavenJArray"/>.</param>
        /// <param name="jsonSerializer">The <see cref="JsonSerializer"/> that will be used to read the object.</param>
        /// <returns>A <see cref="RavenJArray"/> with the values of the specified object</returns>
        public static new RavenJObject FromObject(object o, JsonSerializer jsonSerializer)
        {
            RavenJToken token = FromObjectInternal(o, jsonSerializer);

            if (token != null && token.Type != JTokenType.Object)
                throw new ArgumentException("Object serialized to {0}. JObject instance expected.".FormatWith(CultureInfo.InvariantCulture, token.Type));

            return (RavenJObject)token;
        }

        /// <summary>
        /// Gets or sets the <see cref="RavenJToken"/> with the specified property name.
        /// </summary>
        /// <value></value>
        public RavenJToken this[string propertyName]
        {
            get
            {
                ValidationUtils.ArgumentNotNull(propertyName, "propertyName");
                return Properties[propertyName];
            }
            set { Properties[propertyName] = value; }
        }

        /// <summary>
        /// Loads an <see cref="RavenJObject"/> from a <see cref="JsonReader"/>. 
        /// </summary>
        /// <param name="reader">A <see cref="JsonReader"/> that will be read for the content of the <see cref="RavenJObject"/>.</param>
        /// <returns>A <see cref="RavenJObject"/> that contains the JSON that was read from the specified <see cref="JsonReader"/>.</returns>
        public static RavenJObject Load(JsonReader reader)
        {
            ValidationUtils.ArgumentNotNull(reader, "reader");

            if (reader.TokenType == JsonToken.None)
            {
                if (!reader.Read())
                    throw new Exception("Error reading JObject from JsonReader.");
            }

            if (reader.TokenType != JsonToken.StartObject)
                throw new Exception(
                    "Error reading JObject from JsonReader. Current JsonReader item is not an object: {0}".FormatWith(
                        CultureInfo.InvariantCulture, reader.TokenType));

            if (reader.Read() == false)
                throw new Exception("Unexpected end of json object");

            string propName = null;
            var o = new RavenJObject();
            do
            {
                switch (reader.TokenType)
                {
                    case JsonToken.Comment:
                        // ignore comments
                        break;
                    case JsonToken.PropertyName:
                        propName = reader.Value.ToString();
                        break;
                    case JsonToken.EndObject:
                        return o;
                    case JsonToken.StartObject:
                        if (!string.IsNullOrEmpty(propName))
                        {
                            var val = RavenJObject.Load(reader);
                            o.Properties.Add(propName, val);
                            propName = null;
                        }
                        else
                        {
                            throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
                                                                    .FormatWith(CultureInfo.InvariantCulture,
                                                                                reader.TokenType));
                        }
                        break;
                    case JsonToken.StartArray:
                        if (!string.IsNullOrEmpty(propName))
                        {
                            var val = RavenJArray.Load(reader);
                            o.Properties.Add(propName, val);
                            propName = null;
                        }
                        else
                        {
                            throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
                                                                    .FormatWith(CultureInfo.InvariantCulture,
                                                                                reader.TokenType));
                        }
                        break;
                    default:
                        if (!string.IsNullOrEmpty(propName))
                        {
                            var val = RavenJValue.Load(reader);
                            o.Properties.Add(propName, val);
                            propName = null;
                        }
                        else
                        {
                            throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
                                                                    .FormatWith(CultureInfo.InvariantCulture,
                                                                                reader.TokenType));
                        }
                        break;
                }
            } while (reader.Read());

            throw new Exception("Error reading JObject from JsonReader.");
        }
    }
}

File: Raven.Client.Java/lib/src/main/java/raven/abstractions/json/linq/RavenJObject.java
package raven.abstractions.json.linq;

import java.io.IOException;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.Map.Entry;

import org.apache.commons.lang.StringUtils;
import org.codehaus.jackson.JsonGenerator;
import org.codehaus.jackson.JsonParser;
import org.codehaus.jackson.JsonToken;
import org.codehaus.jackson.map.ObjectMapper;

import raven.abstractions.exceptions.JsonReaderException;
import raven.abstractions.exceptions.JsonWriterException;
import raven.abstractions.extensions.JsonExtensions;

public class RavenJObject extends RavenJToken implements Iterable<Entry<String, RavenJToken>> {

  /**
   * Creates a {@link RavenJObject} from an object.
   * @param o The object that will be used to create {@link RavenJObject}
   * @return A {@link RavenJObject} with the values of the specified object.
   */
  public static RavenJObject fromObject(Object o) {
    return fromObject(o, JsonExtensions.getDefaultObjectMapper());
  }

  /**
   * Creates a {@link RavenJObject} from an object.
   * @param o The object that will be used to create {@link RavenJObject}
   * @param objectMapper The {@link ObjectMapper} that will be used to read the object.
   * @return {@link RavenJObject} with the values of the specified object.
   */
  private static RavenJObject fromObject(Object o, ObjectMapper objectMapper) {
    RavenJToken token = fromObjectInternal(o, objectMapper);

    if (token != null && token.getType() != JTokenType.OBJECT)
      throw new IllegalArgumentException("Object serialized to " + token.getType() + ". RavenJObject instance expected.");

    return (RavenJObject) token;
  }

  public static RavenJObject load(JsonParser parser) {
    try {
      if (parser.getCurrentToken() == null) {
        if (parser.nextToken() == null) {
          throw new JsonReaderException("Error reading RavenJToken from JsonParser");
        }
      }
      if (parser.getCurrentToken() != JsonToken.START_OBJECT) {
        throw new JsonReaderException("Error reading RavenJObject from JsonParser. Current JsonReader item is not an object: " + parser.getCurrentToken());
      }
      if (parser.nextToken() == null) {
        throw new JsonReaderException("Unexpected end of json object");
      }

      String propName = null;
      RavenJObject o = new RavenJObject();
      do {

        switch (parser.getCurrentToken()) {
        case FIELD_NAME:
          propName = parser.getText();
          break;
        case END_OBJECT:
          return o;
        case START_OBJECT:
          if (StringUtils.isNotEmpty(propName)) {
            RavenJObject val = RavenJObject.load(parser);
            o.set(propName, val);
            propName = null;
          } else {
            throw new JsonReaderException("The JsonReader should not be on a token of type " + parser.getCurrentToken());
          }
          break;
        case START_ARRAY:
          if (StringUtils.isNotEmpty(propName)) {
            RavenJArray val = RavenJArray.load(parser);
            o.set(propName, val);
            propName = null;
          } else {
            throw new JsonReaderException("The JsonReader should not be on a token of type " + parser.getCurrentToken());
          }
          break;
        default:
          if (StringUtils.isNotEmpty(propName)) {
            RavenJValue val = (RavenJValue) RavenJToken.load(parser);
            o.set(propName, val);
            propName = null;
          } else {
            throw new JsonReaderException("The JsonReader should not be on a token of type " + parser.getCurrentToken());
          }
          break;

        }
      } while (parser.nextToken() != null);

      throw new JsonReaderException("Error reading RavenJObject from JsonReader.");

    } catch (IOException e) {
      throw new JsonWriterException(e.getMessage(), e);
    }
  }

  /**
   * Loads {@link RavenJObject} from a string that contains JSON.
   * @param json A {@link String} that contains JSON.
   * @return A {@link RavenJObject} populated from the string that contains JSON.
   */
  public static RavenJObject parse(String json) {
    try {
      JsonParser jsonParser = JsonExtensions.getDefaultJsonFactory().createJsonParser(json);
      return load(jsonParser);
    } catch (IOException e) {
      throw new JsonReaderException(e.getMessage(), e);
    }
  }

  private DictionaryWithParentSnapshot properties = new DictionaryWithParentSnapshot();

  public RavenJObject() {
    // empty by design
  }

  public RavenJObject(DictionaryWithParentSnapshot snapshot) {
    properties = snapshot;
  }

  public RavenJObject(RavenJObject other) {
    properties = new DictionaryWithParentSnapshot();
    for (Map.Entry<String, RavenJToken> kv : other.getProperties().entrySet()) {
      properties.put(kv.getKey(), kv.getValue());
    }
  }

  public void add(String propertyName, RavenJToken token) {
    properties.put(propertyName, token);
  }

  @Override
  protected void addForCloning(String key, RavenJToken token) {
    properties.put(key, token);
  }

  @Override
  public RavenJObject cloneToken() {
    return (RavenJObject) cloneTokenImpl(new RavenJObject());
  }

  public boolean containsKey(String key) {
    return properties.containsKey(key);
  }

  @Override
  public RavenJObject createSnapshot() {
    return new RavenJObject(properties.createSnapshot());
  }

  @Override
  public void ensureCannotBeChangeAndEnableShapshotting() {
    properties.ensureSnapshot();
  }

  public void ensureSnapshot(String msg) {
    properties.ensureSnapshot(msg);
  }

  public RavenJToken get(String propertyName) {
    return properties.get(propertyName);
  }

  public int getCount() {
    return properties.size();
  }

  public Set<String> getKeys() {
    return properties.keySet();
  }

  /**
   * @return the properties
   */
  public DictionaryWithParentSnapshot getProperties() {
    return properties;
  }

  @Override
  public JTokenType getType() {
    return JTokenType.OBJECT;
  }

  @Override
  public boolean isSnapshot() {
    return properties.isSnapshot();
  }

  public boolean remove(String propertyName) {
    return properties.remove(propertyName) != null;
  }

  public void set(String propertyName, RavenJToken value) {
    properties.put(propertyName, value);
  }

  /**
   * Writes this token to a {@link JsonGenerator}
   * @param writer A {@link JsonGenerator} into which this method will write.
   */
  @Override
  public void writeTo(JsonGenerator writer) {
    try {
      writer.writeStartObject();

      if (properties != null) {
        for (String key : properties.keySet()) {
          writer.writeFieldName(key);
          RavenJToken value = properties.get(key);
          if (value == null) {
            writer.writeNull();
          } else {
            value.writeTo(writer);
          }
        }
      }
      writer.writeEndObject();
    } catch (IOException e) {
      throw new JsonWriterException(e.getMessage(), e);
    }

  }

  /* (non-Javadoc)
   * @see raven.client.json.RavenJToken#value(java.lang.Class, java.lang.String)
   */
  @SuppressWarnings("unchecked")
  @Override
  public <T> T value(Class<T> clazz, String key) {
    //TODO: finish me!
    // externalize converter

    if (!containsKey(key)) {
      throw new IllegalArgumentException("Object does not contain key: " + key);
    }
    RavenJToken ravenJToken = get(key);
    if (ravenJToken != null) {
      RavenJValue ravenValue = (RavenJValue) ravenJToken;
      switch (ravenJToken.getType()) {
      case STRING:
        if (String.class.isAssignableFrom(clazz)) {
          return (T) ((RavenJValue) ravenJToken).getValue();
        } else if (UUID.class.isAssignableFrom(clazz)) {
          return (T) UUID.fromString((String) ravenValue.getValue());
        }
        break;
      case BOOLEAN:
        if (Boolean.class.isAssignableFrom(clazz)) {
          return (T) ravenValue.getValue();
        }
      }
    }
    throw new IllegalArgumentException("Unsupported conversion. From:" + ravenJToken.getType() + " to " + clazz.getCanonicalName());
  }

  @Override
  public Iterator<Entry<String, RavenJToken>> iterator() {
    return properties.iterator();
  }

}
